tests/avro/test_avro_producer.py

```diff
@@ -20,57 +20,66 @@
+    def test_produce_no_key(self):
+        value_schema = avro.load(os.path.join(avsc_dir, "basic_schema.avsc"))
+        producer = AvroProducer({'schema.registry.url': 'http://127.0.0.1:9001'}, default_value_schema=value_schema)
         try:
-            producer.produce(topic='test', value={"name": 'abc"'}, value_schema=valueSchema, key='mykey')
-            self.fail("Should expect key_schema")
-        except Exception as e:
+            # No key should work
+            producer.produce(topic='test', value={"name": 'abc"'})
+        except SerializerError:
+            self.fail("Produce should accept empty keys")
+        except BaseException:
             pass
 
-    def test_produce_arguments(self):
-        value_schema = avro.load(os.path.join(avsc_dir, "basic_schema.avsc"))
-        producer = AvroProducer({'schema.registry.url': 'http://127.0.0.1:0'}, default_value_schema=value_schema)
-
-        try:
+    def test_produce_no_value_schema(self):
+        producer = AvroProducer({'schema.registry.url': 'http://127.0.0.1:9001'})
+        with self.assertRaises(SerializerError):
+            # Producer should not accept a value with no schema
             producer.produce(topic='test', value={"name": 'abc"'})
-        except Exception as e:
-            exc_type, exc_obj, exc_tb = sys.exc_info()
-            if exc_type.__name__ == 'SerializerError':
-                self.fail()
 
-    def test_produce_arguments_list(self):
-        producer = AvroProducer({'schema.registry.url': 'http://127.0.0.1:0'})
+    def test_produce_no_key_schema(self):
+        value_schema = avro.load(os.path.join(avsc_dir, "basic_schema.avsc"))
+        producer = AvroProducer({'schema.registry.url': 'http://127.0.0.1:9001'}, default_value_schema=value_schema)
         try:
-            producer.produce(topic='test', value={"name": 'abc"'}, key='mykey')
-        except Exception as e:
-            exc_type, exc_obj, exc_tb = sys.exc_info()
-            if exc_type.__name__ == 'SerializerError':
-                pass
+            producer.produce(topic='test', value={"name": 'abc"'}, key={"name": 'abc"'})
+        except SerializerError:
+            pass
+        except BaseException:
+            self.fail("If the key is provided as a dict an avro schema must also be provided")
 
+    def test_produce_value_and_key_schemas(self):
+        value_schema = avro.load(os.path.join(avsc_dir, "basic_schema.avsc"))
+        producer = AvroProducer({'schema.registry.url': 'http://127.0.0.1:9001'}, default_value_schema=value_schema, default_key_schema=value_schema)
+        try:
+            producer.produce(topic='test', value={"name": 'abc"'}, key={"name": 'abc"'})
+        except SerializerError:
+            self.fail("Produce should accept both value and keys based on avro")
+        except BaseException:
+            pass
 
-def suite():
-    return unittest.TestLoader().loadTestsFromTestCase(TestAvroProducer)
+    def test_produce_string_key(self):
```

Comment:
And now I think I see what you were getting at. Isn't this not really safe? You're basically skipping serialization for everything that isn't a dict, but I think what you were actually trying to get is that serialization works as a sort of pass-through restrict to `str` for py2 (assuming you can dump as simple bytes) and `bytes` for py3 for the keys. These are the only translations that make sense since everything else requires *some* sort of serialization which Avro provides (even simple strings, which I would agree should 99.999% of the time be serialized as UTF-8, but that shouldn't be absolutely assumed).

This is actually pretty tricky. Avro also provides serialization for pure bytes where that is the entirety of the schema. In some cases it is useful to be able to wrap those raw bytes. The Confluent serializers do assume that if you pass bytes, you're going to be passing raw bytes for everything and omits the framing. So we could do that for a compatible implementation, but we'd want to make sure the passthrough *only* works for str/bytes. Also, this isn't handled differently between keys and values, so the implementations should really be consistent. (Again this ties back to the fact that the current implementation is only supporting a subset of types for serialization.)

(Note, also, that this comes at a cost. It makes it impossible to mix and match types within the same topic because once you drop raw bytes in one message without any framing, there's no way to distinguish the different schemas.)

By the way, if you're thinking about making primitive types work more generally (i.e. more than just bytes, even if just utf-8 strings), I think there are basically 2 ways of making this happen. The first would be to adjust the way AvroProducer/Consumer are implemented, instead providing a SerializingProducer/Consumer and passing in the desired serializers. This lets you control the key and value serialization formats independently. Actually, looking at this now @criccomini, it seems like that might be a better path for you folks given requests re: the REST proxy? The other would be to allow a "primitive" mode in this class that knows how to serialize primitive types outside of Avro, handling the trivial types like string (utf-8), int (32-bit network order int encoding, although that gets tricker with the unification of int/long in py3), and long (64-bit network order int).

I'm probably more in favor of the former since the latter would just result in multiple implementations (albeit trivial) across different XProducers for "X" serialization format. It also gives you an easy way to provide one set of documentation for everyone with the understanding that they just need to understand the serialization library's type support for complex types. Opinions?